[DOC] Keycloak example edit (#5287)
strimzi/strimzi-kafka-operator: documentation/modules/oauth/proc-oauth-authorization-keycloak-example.adoc

Grants are cached and enforced locally on the Kafka broker for each user session to provide fast authorization decisions.

Strimzi provides two example files with the deployment artifacts for setting up Keycloak:

`kafka-ephemeral-oauth-single-keycloak-authz.yaml`:: An example `Kafka` custom resource configured for OAuth 2.0 token-based authorization using Keycloak.
You can use the custom resource to deploy a Kafka cluster that uses `keycloak` authorization and token-based `oauth` authentication.

`kafka-authz-realm.json`:: An example Keycloak realm configured with sample groups, users, roles and clients.
You can import the realm into a Keycloak instance to set up fine-grained permissions to access Kafka.

If you want to try Keycloak, use these files to perform the tasks  outlined in this section in the order shown.

. xref:proc-oauth-authorization-keycloak-example-setup_{context}[Setting up the Keycloak Admin Console]
. xref:proc-oauth-authorization-keycloak-example-deploy-kafka_{context}[Deploying a Kafka cluster with Keycloak authorization]
. xref:proc-oauth-authorization-keycloak-example-authentication_{context}[Authenticating Keycloak access to Kafka brokers]

> . xref:proc-oauth-authorization-keycloak-example-check_{context}[Checking authorized access to Kafka using permissions]

Reviewer: ```suggestion
. xref:proc-oauth-authorization-keycloak-example-check_{context}[Checking authorized access to Kafka using a CLI Kafka client session]
```

.Security considerations

> Keycloak authorization uses TLS connectivity where necessary. The server certificates are auto-generated. TLS termination for the Keycloak server is on a reverse proxy server using a Kubernetes ingress, not the Keycloak server pod. Take care when creating and reviewing any production configuration.

Author: I rewrote based on your explanation. See above

> Keycloak requires Kafka clients to use the same hostname and port as the Kafka brokers. This example assumes that the ingress hostname is visible to the pods inside the Kubernetes cluster. In your cluster, the ingress might be visible to external Kafka clients, but not to the pods. You can deploy the Keycloak server to a different Kubernetes cluster than the Cluster Operator.

Author: @mstruk I'm not sure I fully understand what we're trying to say with this section. Can we reduce it down and pull out the key point? For example, "In a production environment, you can deploy...to avoid..."

Author: Thanks for that, Marko. It made it a lot clearer. I've updated as follows:

**Security considerations**
Kafka and Kafka clients must connect to Keycloak for authorization to work. Connection to Keycloak is indirect. The connection routes through an ingress HTTP gateway using a hostname on port 443. Kafka clients use the same hostname and port as the Kafka brokers. The gateway is part of the Kubernetes infrastructure. This example assumes that Keycloak is deployed using the Keycloak Operator. The operator makes the gateway hostname visible to pods in the same Kubernetes cluster. The hostname must be visible for the example to work. If it's not visible, try deploying Keycloak to a separate Kubernetes cluster. The pods can then access Keycloak using the external hostname to make the connection.

Author: We could add a graphic if it helps

Reviewer: I think it still needs a little bit of clarifying. This statement is incorrect:
'The operator makes the gateway hostname visible to pods in the same Kubernetes cluster.' 
Keycloak Operator can't make the gateway hostname visible to pods, it can only make use of the ingress. It can't ensure that the ingress hostname is visible to the pods - that part is a matter of Kubernetes administrator configuring the Kubernetes cluster to make the ingress hostname visible to the pods within that cluster.
'The operator exposes Keycloak over TLS using the ingress HTTP gateway. The ingress hostname must be visible for the example to work'. I don't think graphics is necessary, but if you think it is, sure.

+
[source,shell,subs="+quotes"]
----
SSO_HOST=_SSO-HOSTNAME_
SSO_HOST_PORT=$SSO_HOST:443
STOREPASS=storepass

echo "Q" | openssl s_client -showcerts -connect $SSO_HOST_PORT 2>/dev/null | awk ' /BEGIN CERTIFICATE/,/END CERTIFICATE/ { print $0 } ' > /tmp/sso.crt
----
+
The certificate is required as Kubernetes ingress is used to make a secure (HTTPS) connection.

. Deploy the certificate to Kubernetes as a secret.
+
[source,shell]
----
kubectl create secret generic oauth-server-cert --from-file=/tmp/sso.crt -n $NS
----

. Set the hostname as an environment variable
+
[source,shell,subs="+quotes"]
----
SSO_HOST=_SSO-HOSTNAME_
EXAMPLES_DIR=examples
cat $EXAMPLES_DIR/security/keycloak-authorization/kafka-ephemeral-oauth-single-keycloak-authz.yaml | sed -E 's#\${SSO_HOST}'"#$SSO_HOST#" | kubectl create -n $NS -f -
----

. Create and deploy the example Kafka cluster.
+
[source,shell]
----
kubectl apply -f examples/security/keycloak-authorization/kafka-ephemeral-oauth-single-keycloak-authz.yaml
----

[id='proc-oauth-authorization-keycloak-example-authentication_{context}']

> == Authenticating Keycloak access to Kafka brokers

Author: @mstruk Is this title accurate? Are teh steps split correctly in this procedure?

----
SSO_HOST=_SSO-HOSTNAME_
SSO_HOST_PORT=$SSO_HOST:443
STOREPASS=storepass

echo "Q" | openssl s_client -showcerts -connect $SSO_HOST_PORT 2>/dev/null | awk ' /BEGIN CERTIFICATE/,/END CERTIFICATE/ { print $0 } ' > /tmp/sso.crt
----

. Create a truststore for TLS connection to the Kafka brokers.
+
[source,shell,subs="+quotes"]
----
keytool -keystore /tmp/truststore.p12 -storetype pkcs12 -alias sso -storepass $STOREPASS -import -file /tmp/sso.crt -noprompt
----

. Use the Kafka bootstrap address as the hostname of the Kafka broker and the `tls` listener port (9093) to prepare a certificate for the Kafka broker.
+
[source,shell]
----
KAFKA_HOST_PORT=my-cluster-kafka-bootstrap:9093
STOREPASS=storepass

echo "Q" | openssl s_client -showcerts -connect $KAFKA_HOST_PORT 2>/dev/null | awk ' /BEGIN CERTIFICATE/,/END CERTIFICATE/ { print $0 } ' > /tmp/my-cluster-kafka.crt
----

. Add the certificate for the Kafka broker to the truststore.
+
[source,shell]
----
keytool -keystore /tmp/truststore.p12 -storetype pkcs12 -alias my-cluster-kafka -storepass $STOREPASS -import -file /tmp/my-cluster-kafka.crt -noprompt
----

[id='proc-oauth-authorization-keycloak-example-check_{context}']

> == Checking authorized access to Kafka using permissions

Reviewer: ```suggestion
== Checking authorized access to Kafka using a CLI Kafka client session
```

> == Checking authorized access to Kafka using permissions

Author: Thanks. I prefer the split. Maybe someone doesn't want to set up the example, but can look at how it works in the checks.

Check the authorization rules applied through the Keycloak realm.
Apply the checks using Kafka's example producer and consumer clients to create topics with user and service accounts that have different levels of access.

Use the `team-a-client` and `team-b-client` clients to check the authorization rules.
Use the `alice` admin user to perform additional administrative tasks on Kafka.

.Prerequisites

* ZooKeeper and Kafka are running in the Kubernetes cluster to be able to send and receive messages.

> * Access between the Kafka brokers and Keycloak is authenticated.

Reviewer: ```suggestion
* The steps in previous chapter have been performed and the same interactive CLI Kafka client session is continued from the point left in the previous chapter.
```

Author: Mentioned to keep the session open in the previous task.

> * The latest Kafka producer and consumer binaries are installed on your local machine.

Reviewer: In this example we're using a specific released version of Strimzi Kafka image which already contains the latest binaries. 
```suggestion
```

----

. Create a truststore for TLS connection to the Kafka brokers.
+
[source,shell,subs="+quotes"]
----
keytool -keystore /tmp/truststore.p12 -storetype pkcs12 -alias sso -storepass $STOREPASS -import -file /tmp/sso.crt -noprompt
----

. Use the Kafka bootstrap address as the hostname of the Kafka broker and the `tls` listener port (9093) to prepare a certificate for the Kafka broker.
+
[source,shell]
----
KAFKA_HOST_PORT=my-cluster-kafka-bootstrap:9093
STOREPASS=storepass

echo "Q" | openssl s_client -showcerts -connect $KAFKA_HOST_PORT 2>/dev/null | awk ' /BEGIN CERTIFICATE/,/END CERTIFICATE/ { print $0 } ' > /tmp/my-cluster-kafka.crt
----

. Add the certificate for the Kafka broker to the truststore.
+
[source,shell]
----
keytool -keystore /tmp/truststore.p12 -storetype pkcs12 -alias my-cluster-kafka -storepass $STOREPASS -import -file /tmp/my-cluster-kafka.crt -noprompt
----
+
Keep the session open to check authorized access.

[id='proc-oauth-authorization-keycloak-example-check_{context}']
== Checking authorized access to Kafka using a CLI Kafka client session

Check the authorization rules applied through the Keycloak realm using an interactive CLI session.

> Apply the checks using Kafka's example producer and consumer clients to create topics with user and service accounts that have different levels of access.

Reviewer 2: ```suggestion
Perform the checks using Kafka's example producer and consumer clients to create topics with user and service accounts that have different levels of access.
```

* The xref:proc-oauth-authorization-keycloak-example-authentication_{context}[interactive CLI Kafka client session] is started.
+
{ApacheKafkaDownload}.

.Setting up client and admin user configuration

. Prepare a Kafka configuration file with authentication properties for the `team-a-client` client.
+
[source,shell,subs="+quotes"]
----
SSO_HOST=_SSO-HOSTNAME_

cat > /tmp/team-a-client.properties << EOF
security.protocol=SASL_SSL
ssl.truststore.location=/tmp/truststore.p12
ssl.truststore.password=$STOREPASS
ssl.truststore.type=PKCS12
sasl.mechanism=OAUTHBEARER
sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required \
  oauth.client.id="team-a-client" \
  oauth.client.secret="team-a-client-secret" \
  oauth.ssl.truststore.location="/tmp/truststore.p12" \
  oauth.ssl.truststore.password="$STOREPASS" \
  oauth.ssl.truststore.type="PKCS12" \
  oauth.token.endpoint.uri="https://$SSO_HOST/auth/realms/kafka-authz/protocol/openid-connect/token" ;
sasl.login.callback.handler.class=io.strimzi.kafka.oauth.client.JaasClientOauthLoginCallbackHandler
EOF
----
+

> The SASL/OAUTHBEARER mechanism is used.

Reviewer 2: Referred to elsewhere as the OAUTHBEARER mechanism (although there are some exceptions).

The client authenticates with the refresh token that was authenticated in the previous step.
The refresh token requests an access token behind the scenes, which is then sent to the Kafka broker for authentication.

.Producing messages with authorized access

Use the `team-a-client` configuration to check that you can produce messages to topics that start with `a_` or `x_`.

. Write to topic `my-topic`.
+
[source,shell]
----
bin/kafka-console-producer.sh --broker-list my-cluster-kafka-bootstrap:9093 --topic my-topic \
  --producer.config=/tmp/team-a-client.properties
First message
----
+
This request returns a `Not authorized to access topics: [my-topic]` error.
+
`team-a-client` has a `Dev Team A` role that gives it permission to perform any supported actions on topics that start with `a_`, but can only write to topics that start with `x_`.
The topic named `my-topic` matches neither of those rules.

. Write to topic `a_messages`.
+
[source,shell]
----
bin/kafka-console-producer.sh --broker-list my-cluster-kafka-bootstrap:9093 --topic a_messages \
  --producer.config /tmp/team-a-client.properties
First message
Second message
----
+
Messages are produced to Kafka successfully.

> . Press CTRL-C to exit the CLI application.

Reviewer 2: ```suggestion
. Press CTRL+C to exit the CLI application.
```

. Check the Kafka container log for a debug log of `Authorization GRANTED` for the request.

[source,shell,subs="attributes"]
----
kubectl logs my-cluster-kafka-0 -f -n $NS
----

.Consuming messages with authorized access

Use the `team-a-client` configuration to consume messages from topic `a_messages`.

. Fetch messages from topic `a_messages`.
+
[source,shell,subs=+quotes]
----
bin/kafka-console-consumer.sh --bootstrap-server my-cluster-kafka-bootstrap:9093 --topic a_messages \
  --from-beginning --consumer.config /tmp/team-a-client.properties
----
+

> The request returns an error as the `Dev Team A` role for `team-a-client` only has access to consumer groups that have names starting with `a_`.

Reviewer 2: ```suggestion
The request returns an error because the `Dev Team A` role for `team-a-client` only has access to consumer groups that have names starting with `a_`.
```

. xref:proc-oauth-authorization-keycloak-example-setup_{context}[Accessing the Keycloak Admin Console]
. xref:proc-oauth-authorization-keycloak-example-deploy-kafka_{context}[Deploying a Kafka cluster with Keycloak authorization]
. xref:proc-oauth-authorization-keycloak-example-authentication_{context}[Preparing TLS connectivity for a CLI Kafka client session]
. xref:proc-oauth-authorization-keycloak-example-check_{context}[Checking authorized access to Kafka using a CLI Kafka client session]

.Connection to Keycloak
Kafka and Kafka clients must connect to Keycloak for authorization to work. Connection to Keycloak is indirect.
The connection routes through an ingress HTTP gateway using a hostname on port 443. Kafka clients use the same hostname and port as the Kafka brokers.
The gateway is part of the Kubernetes infrastructure. This example assumes that Keycloak is deployed using the Keycloak Operator.
The Keycloak Operator exposes Keycloak over TLS using the ingress HTTP gateway.
The ingress hostname must be visible to the pods within the cluster for the example to work.
If it's not visible, try deploying Keycloak to a separate Kubernetes cluster.

> The pods can then access Keycloak using the external hostname to make the connection.

Reviewer 3: Feel free to postpone it to separate PR if it was there already before - from the PR it is not clear if the content just moved or is new. 

But I do not think this makes sense. If I run Kafka cluster and clients only inside the same Kubernetes cluster, why would I need to expose Keycloak externally or deploy it in a separate Kubernetes cluster? I think this should just say something like ...

```
The Kafka brokers need to use the same address to access the Keycloak server as the Kafka clients.
For example, if the Kafka clients use an external address based on an Ingress or Route, the Kafka broker should use the same.
```

However, I'm not sure how this applies to authorization. I think it applies only to authentication. So it probably should not be here at all because it will confuse things: `authorization` is configured only once per cluster and needs to use the same address for all clients. `authentication` is configured per-listener and might need to use completely different address for each listener. So maybe we should just remove it from here?

Author: @scholzj Thanks. It's been rewritten since the last publication. It's to explain how Keycloak Operator exposes Keycloak before we get into the procedures, as we're assuming Keycloak is deployed that way. 
@mstruk Hi Marko. Jakub suggests we reduce this Keycloak connection section.

Reviewer: @scholzj It applies to authorizazion as well since the 'keycloak' authorization has to use the same token endpoint to retrieve a list of grants for the token, and access it in the same way it was accessed when obtaining the token.

But maybe the text could be slightly specific to authorization:

Keycloak authorization uses the token that passed authentication validation checks to obtain from Keycloak a list of grants for the user. Thus, it needs to connect to Keycloak, and it needs to use the same hostname and port that were used when obtaining the token for the authentication. See the `Connection to Keycloak` chapter for 'oauth' authentication about Keycloak Ingres hostname visibility to pods in the cluster.

. xref:proc-oauth-authorization-keycloak-example-authentication_{context}[Preparing TLS connectivity for a CLI Kafka client session]
. xref:proc-oauth-authorization-keycloak-example-check_{context}[Checking authorized access to Kafka using a CLI Kafka client session]

.Authentication
When you configure token-based `oauth` authentication, you specify a `jwksEndpointUri` as the URI for local JWT validation.
When you configure `keycloak` authorization, you specify a `tokenEndpointUri` as the URI of the Keycloak token endpoint.

> The hostname for both URIs must be the same.

Author: @mstruk @scholzj replaces the old note
